blocks/autocompleters/index.js

```diff
@@ -51,55 +63,52 @@ import BlockIcon from '../block-icon';
  * @typedef {Object} Completer
  * @property {?String} className A class to apply to the popup menu.
  * @property {String} triggerPrefix the prefix that will display the menu.
- * @property {FnGetOptions} getOptions get the block options in a resolved promise.
+ * @property {FnGetOptions} getOptions get the completer options in a resolved promise.
+ * @property {?FnGetOptionKeywords} getOptionKeywords get the keywords for a given option.
+ * @property {FnRenderOptionLabel} renderOptionLabel render the label for a given option.
  * @property {?FnAllowNode} allowNode filter the allowed text nodes in the autocomplete.
  * @property {?FnAllowContext} allowContext filter the context under which the autocomplete activates.
- * @property {FnOnSelect} onSelect
+ * @property {FnGetOptionCompletion} getOptionCompletion get the completion associated with a given option.
  */
 
 /**
  * Returns an "completer" definition for selecting from available blocks to replace the current one.
  * The definition can be understood by the Autocomplete component.
  *
- * @param {Function} onReplace  Callback to replace the current block.
- *
  * @return {Completer} Completer object used by the Autocomplete component.
  */
-export function blockAutocompleter( { onReplace } ) {
-	// Prioritize common category in block type options
-	const options = sortBy(
-		getBlockTypes(),
-		( { category } ) => 'common' !== category
-	).map( ( blockType ) => {
-		const { name, title, icon, keywords = [] } = blockType;
-		return {
-			value: name,
-			label: [
-				<BlockIcon key="icon" icon={ icon } />,
-				title,
-			],
-			keywords: [ ...keywords, title ],
-		};
-	} );
-
-	const getOptions = () => Promise.resolve( options );
+export function blockAutocompleter() {
+	const getOptions = once( () => Promise.resolve(
+		// Prioritize common category in block type options
+		sortBy(
+			getBlockTypes(),
+			( { category } ) => 'common' !== category
+		)
+	) );
+	const getOptionKeywords = ( { title, keywords = [] } ) => [ ...keywords, title ];
+	const renderOptionLabel = ( { icon, title } ) => [
+		<BlockIcon key="icon" icon={ icon } />,
+		title,
+	];
 
 	const allowContext = ( before, after ) => {
 		return ! ( /\S/.test( before.toString() ) || /\S/.test( after.toString() ) );
 	};
 
-	const onSelect = ( blockName ) => {
-		onReplace( createBlock( blockName ) );
-	};
```

Comment:
Removing this creates a need to check that the result is a block outside of the autocompleter. Imagine an autocompleter that returns a block but we don't way to replace, we want to append something.

Three options:

 - keep onSelect
 - add and `action` property that could have things like `replace`, `append`, `insert-at-caret` values...

Reply:
Today, the block completer has a runtime dependency on a block's `onReplace`. The paragraph block provides `onReplace` when it creates the block completer, but it's less clear to me how to satisfy such dependencies while allowing completers to be overridden during block registration.

I started thinking that maybe completers should simply provide completions without imperative logic regarding what to do with a completion. It's better to keep things simpler and less entangled, but I see a couple of problems with my thinking:

1. You make a good point that we may want to do other things with completions, like append. I'm somewhat attracted to the option of tagging completers with actions like 'insert-at-caret', 'replace', and 'append' but am thinking it's probably a mistake to constrain completers to declarative options.
2. For the block completer to include reusable blocks (see #3791), I believe we need to satisfy a runtime dependency on the editor state, so removing `onSelect` does not directly address the question of completer dependencies. My thinking for this was to add a `blocks.registerBlockType` filter to override the block completer's `getOptions` with a version based on the editor store. This is hacky IMO.

@youknowriad Do you have any thoughts on how we could provide runtime dependencies to completers in a cleaner, more generic way?

Reply:
> My thinking for this was to add a blocks.registerBlockType filter to override the block completer's getOptions with a version based on the editor store. This is hacky IMO.

Actually, I don't think it's hacky, it means that if you use the blocks module only, you get the generic blocks, if you use the editor blocks you get the generic  + reustable blocks. I think this goes in concert with the need to move the reusable blocks definition to the editor module as well. cc @noisysocks 

The runtime dependency we need to satisfy here `onReplace` is different though because it's not a global dependency, it's contextual to the block. and I'm fine with tagging completers to solve this => this means the generic autocompleter behavior is block-aware but it's fine I think.